README.md
# topology-collector

A repository for code that collects the following information about interfaces on compute hosts:

1. Neighbor LLDP connectivity information
2. VF port PCI information for the interface

The output of the code is a JSON report.

## Summary instructions

1. Clone the repo to the OpenStack controller node (requires git)
2. Install Ansible 2.1+ on the OpenStack Controller node
3. Configure passwordless ssh from the OpenStack Conroller node to all compute nodes
4. Update local variables (see below)
5. Execute `ansible-playbook -i controllers get_computes.yml` (Skip if you hand-edit `hypervisors` or you have already run this step previously and no changes in the list of compute nodes are required.)
6. Execute `ansible-playbook -i hypervisors get_topo.yml`

## Details

### Assumptions
1. The interfaces to be processed on the compute nodes are currently `UP` as reported by the command `ip addr`
2. The VFs for the interfaces to be processed on the compute nodes are listed in the directory /sys/class/net/<interface>/device/virt* on each compute node

### Input variable files

Input variables are contained in the following files:
1. `user_vars.yml`
2. `controllers`
3. `hypervisors`

#### `user_vars.yml`

- `temp_dir`, the location on the OpenStack controller node where intermediate files are written to and read from
- `output_dir`, the location on the OpenStack controller node where the date-stamped output files will be written
- `output_file_prefix`, text to prepend to the output file name, e.g. <output_file_prefix>.<date>@<time>.json.
- `interface_regex`, regex to match interface names on the compute nodes. Default is `['*']`

#### `controllers`

Contains a flat list of controller node host names or IP addresses under the tag `[controllers]`. In most cases, this will be a list of one and only one controller. In addition, each controller must have the osc_env_file path provided, e.g.

```
[controllers]
controller_hostname osc_env_file=/path/to/env/file/to-source
```

Note that osc_env_file is the file you would source prior to executing `nova` commands. osc_env_file *must* be included for each controller in the list.


#### `hypervisors`

Contains a flat list of compute node host names or IP addresses under the tag `[computes]`. This list may be populated manually or automatically. Automatic population is achieved using the get-computes role. Each hypervisor must contain both the hypervisor host name and the service_host name as shown in `nova hypervisor-show`.

```
[hypervisors]
hypervisor_hostname service_host=service_host_name
```

### `get-computes`

The `get-computes` role queries the OpenStack controller nodes for the list of nova hypervisors. It parses the output and writes the result to `./hypervisors`.

### `get_topo.yml`

The `get_topo.yml` playbook is the main playbook for gathering topology information and producing the JSON report. It cleans up temporary files from previous runs, makes sure the temp directory exists, then executes the interface and report roles. These roles have their own playbooks and may be executed individually. The other playbooks, each representing an Ansible role, are:

- `topology.yml`, when executed, queries each hosts listed in the `hypervisors` file's `hypervisors` group. The output of this stage is a set of files, one per compute node, in the `temp_dir`.
- `report.yml`, when executed, pulls in the content of each JSON file found in `temp_dir` and creates a full report for all compute nodes. The full report is written to `output_dir` using a unique file name.

### Custom Ansible modules and filters

The implementation includes the following custom Ansible modules written in Python:

- `library\interfaces.py`, a module to query each compute node for information about its interfaces. Matching interfaces are filtered using `state` and regex match on name.
- `library\topology.py`, a module that executes the commands for collecting infomration about each interface, converting the output to JSON.
- `filter_plugins\nova.py`, a set of filters for massaging the outputs of various nova commands.

### Output

The output of the run will be a file that contains a JSON string. The schema itself and a sample output can be found in the `schema` subdirectory and pasted, below.

The name of the file will be of the form `<output_file_prefix>.<date>@<time>.json`. For example, `blue.2016-11-21@14:16:18.json`.

#### JSON schema

```
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "title": "Nuage Topology Collector Report",
    "description": "A report of 1) the mapping between compute host interfaces and VSG ToR ports and 2) the VF PCI information for the interfaces",
    "type": "object",
    "properties": {
        "datetime": {
            "description": "The date and time this report was generated",
            "type": "string"
        },
        "compute-hosts": {
            "type": "array",
            "items": {
                "$ref": "#/definitions/compute-host"
            }
        }
    },
    "required": ["datetime", "compute-hosts"],
    "definitions": {
        "compute-host": {
            "properties": {
                "hypervisor hostname": {
                    "type": "string",
                    "description": "Hypervisor hostname of the compute host"
                },
                "service_host name": {
                    "type": "string",
                    "description": "service_host name of the compute host"
                },
                "interfaces": {
                    "type": "array",
                    "description": "zero or more interfaces on the compute host",
                    "items": {
                        "$ref": "#/definitions/interface"
                    }
                }
            },
            "required": [ "hypervisor hostname", "service_host name",  "interfaces" ]
        },
        "interface": {
            "properties": {
                "name": {
                    "type": "string",
                    "description": "The name of the interface"
                },
                "vf-info": {
                    "type": "array",
                    "description": "The VF information for the interface",
                    "items": {
                        "$ref": "#/definitions/vf-info"
                    }
                },
                "neighbor-system-name": {
                    "type": "string",
                    "description": "The system name of the connected neighbor"
                },
                "neighbor-system-mgmt-ip": {
                    "type": "string",
                    "description": "The management IP address of the connected neighbor"
                },
                "neighbor-system-port": {
                    "type": "string",
                    "description": "The port designation on the connected neighbor"
                }
            },
            "required": [ "name", "neighbor-system-name", "neighbor-system-port" ]
        },
        "vf-info": {
            "properties": {
                "pci-id": {
                    "type": "string",
                    "description": "The VF PCI ID"
                },
                "device-name": {
                    "type": "string",
                    "description": "The VF device name"
                }
            }
        }
    }
}
```

#### JSON sample

```
{
    "datetime": "2016-12-17@04:06:05",
    "compute-hosts": [
        {
            "hypervisor hostname": "andc-ubuntu02.an.nuagenetworks.net",
            "service_host name": "andc-ubuntu02",
            "interfaces": [

                {
 "name": "eno1",
 "vf_info": [],
 "neighbor-system-name": "cas-sf3-009",
 "neighbor-system-mgmt-ip": "10.101.2.2",
 "neighbor-system-port": "1/1/9"  },

                {
 "name": "eno4",
 "vf_info": [],
 "neighbor-system-name": "None",
 "neighbor-system-mgmt-ip": "None",
 "neighbor-system-port": "None"  },

                {
 "name": "ens15f0",
 "vf_info": [],
 "neighbor-system-name": "cas-sf3-009",
 "neighbor-system-mgmt-ip": "10.101.2.2",
 "neighbor-system-port": "1/1/1"  },

                {
 "name": "ens15f1",
 "vf_info": [
 { "device-name": "virtfn0", "pci-id": "0000:03:06.0" },
 { "device-name": "virtfn1", "pci-id": "0000:03:06.1" },
 { "device-name": "virtfn10", "pci-id": "0000:03:07.2" },
 { "device-name": "virtfn11", "pci-id": "0000:03:07.3" },
 { "device-name": "virtfn12", "pci-id": "0000:03:07.4" },
 { "device-name": "virtfn13", "pci-id": "0000:03:07.5" },
 { "device-name": "virtfn14", "pci-id": "0000:03:07.6" },
 { "device-name": "virtfn15", "pci-id": "0000:03:07.7" },
 { "device-name": "virtfn16", "pci-id": "0000:03:08.0" },
 { "device-name": "virtfn17", "pci-id": "0000:03:08.1" },
 { "device-name": "virtfn18", "pci-id": "0000:03:08.2" },
 { "device-name": "virtfn19", "pci-id": "0000:03:08.3" },
 { "device-name": "virtfn2", "pci-id": "0000:03:06.2" },
 { "device-name": "virtfn20", "pci-id": "0000:03:08.4" },
 { "device-name": "virtfn21", "pci-id": "0000:03:08.5" },
 { "device-name": "virtfn22", "pci-id": "0000:03:08.6" },
 { "device-name": "virtfn23", "pci-id": "0000:03:08.7" },
 { "device-name": "virtfn24", "pci-id": "0000:03:09.0" },
 { "device-name": "virtfn25", "pci-id": "0000:03:09.1" },
 { "device-name": "virtfn26", "pci-id": "0000:03:09.2" },
 { "device-name": "virtfn27", "pci-id": "0000:03:09.3" },
 { "device-name": "virtfn28", "pci-id": "0000:03:09.4" },
 { "device-name": "virtfn29", "pci-id": "0000:03:09.5" },
 { "device-name": "virtfn3", "pci-id": "0000:03:06.3" },
 { "device-name": "virtfn30", "pci-id": "0000:03:09.6" },
 { "device-name": "virtfn31", "pci-id": "0000:03:09.7" },
 { "device-name": "virtfn4", "pci-id": "0000:03:06.4" },
 { "device-name": "virtfn5", "pci-id": "0000:03:06.5" },
 { "device-name": "virtfn6", "pci-id": "0000:03:06.6" },
 { "device-name": "virtfn7", "pci-id": "0000:03:06.7" },
 { "device-name": "virtfn8", "pci-id": "0000:03:07.0" },
 { "device-name": "virtfn9", "pci-id": "0000:03:07.1" }],
 "neighbor-system-name": "cas-sf3-009",
 "neighbor-system-mgmt-ip": "10.101.2.2",
 "neighbor-system-port": "1/2/1"  }
            ]
        }
    ]
}
```

## Build & Package installation

Package build scripts are under `build` directory

to build ubuntu packages, just run

```
bash build/build_nuage_deb.sh
```

to build el7 packages, just run
```
build build_nuage_rpm.sh
```

This package requires ansible >= 2.1.0 as dependency

for ubuntu, install from ansible ppa to get latest ansible package.
```
sudo add-apt-repository ppa:ansible/ansible
sudo apt-get update
sudo apt-get install ansible
```

for centos7, install latest ansible from `EPEL`
```
sudo yum install epel-release
sudo yum install ansible
```

then install this package by either
```
sudo dpkg -i nuage-topology-collector*
```
or
```
sudo yum localinstall nuage-topology-collector*
```

the content of this package is under `/opt/nuage/topology-collector`


debian/changelog
nuage-topology-collector (0:0.0.0-0) unstable; urgency=low

  dummy changelog

 -- Nuage Networks <info@nuagenetworks.com>  Wed, 14 Jan 2015 15:21:43 -0800


library/lldp_neighbor_tlv.py
#!/usr/bin/python

import datetime
import re
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.six import b

# Copyright 2016 Nokia
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

DOCUMENTATION = '''
---
module: lldp_neighbor_tlv
short_description: Given an interface, returns a JSON representation of the neighbor's LLDP TLV
options:
  system_name:
    description:
      - The system name or IP address of the compute node being processed
    required: true
  interface:
    description:
      - The network interface to interrogate
    required: true
'''

EXAMPLES = '''
# Verify the state of program "ntpd-status" state.
- lldpneighbortlv: system_name=192.168.1.1 interfaace=eth0
- lldpneighbortlv: system_name=cas-cs2-011 interfaace=eno4
'''

# TODO: Update this description. Rename the function.
# convert_lldptool_output_to_json is a function that takes an output string from a
# specific run of lldptool and converts it to JSON. The specific form of the lldptool
# command is 'lldptool -t -n -i <interface>'. That is the form used to gather the
# interface's neighbor TLV. The code, below, is tied very tightly to the exact
# format of the command output. If that output changes, this function must change.


def convert_lldptool_output_to_json(interface, lldpout, lsout):

    LLDPSYSTEMNAME = "System Name TLV"
    NEIGHBORNAME = "neighbor-system-name"
    LLDPSYSTEMIP = "Management Address TLV"
    NEIGHBORIP = "neighbor-system-mgmt-ip"
    LLDPSYSTEMPORT = "Port Description TLV"
    NEIGHBORPORT = "neighbor-system-port"

    # Insert the interface name into the JSON object.
    parsed = "\n \"name\": \"%s\"" % interface

    # Insert VF port info
    vf_info = ",\n \"vf_info\": ["
    lines = lsout.split('\n')
    found_first_match = False
    for line in lines:
        if re.search(" virt", line):
            fmt = ",\n { \"device-name\": \"%s\""
            vf_parts = line.split(' ')
            if len(vf_parts) >= 17:
                if not found_first_match:
                    found_first_match = True
                    fmt = "\n { \"device-name\": \"%s\""
                vf_info += fmt % vf_parts[16]
            if len(vf_parts) >= 19:
                pci_id_parts = vf_parts[18].split('/')
                if len(pci_id_parts) >= 2:
                    vf_info += ", \"pci-id\": \"%s\" }" % pci_id_parts[1]
    vf_info += "]"
    parsed += vf_info

    # Now add neighbor information
    scratch = lldpout.replace('\n\t', '\t').strip()
    neighborname = "None"
    neighborip = "None"
    neighborport = "None"
    for line in scratch.split('\n'):
        if re.search( LLDPSYSTEMNAME, line):
            sys_name_tlv_parts = line.split('\t')
            if len(sys_name_tlv_parts) >= 2:
                neighborname = sys_name_tlv_parts[1]
        elif re.search( LLDPSYSTEMIP, line):
            mgmt_addr_tlv_parts = line.split('\t')
            if len(mgmt_addr_tlv_parts) >= 2:
                mgmt_addr_parts = mgmt_addr_tlv_parts[1].split(' ')
                if len(mgmt_addr_parts) >= 2:
                    neighborip = mgmt_addr_parts[1]
        elif re.search( LLDPSYSTEMPORT, line):
            port_desc_tlv_parts = line.split('\t')
            if len(port_desc_tlv_parts) >= 2:
                port_desc_parts = port_desc_tlv_parts[1].split(' ')
                if len(port_desc_parts) >= 2:
                    neighborport = port_desc_parts[1]
    parsed += ",\n \"%s\": \"%s\"" % ( NEIGHBORNAME, neighborname)
    parsed += ",\n \"%s\": \"%s\"" % ( NEIGHBORIP, neighborip)
    parsed += ",\n \"%s\": \"%s\" " % ( NEIGHBORPORT, neighborport)
    return "{ %s }" % parsed


def main():
    arg_spec = dict(
        system_name=dict(required=True),
        interface=dict(required=True)
    )

    module = AnsibleModule(argument_spec=arg_spec)

    system_name = module.params['system_name']
    interface = module.params['interface']

    LLDPTOOL = module.get_bin_path('lldptool', True)
    LS = module.get_bin_path('ls', True)

    startd = datetime.datetime.now()

    lldpcmd = "%s -t -n -i %s" % (LLDPTOOL, interface)
    lldprc, lldpout, lldperr = module.run_command(lldpcmd, check_rc=True)

    if lldperr is None:
        lldperr = b('')

    if lldpout is None:
        lldpout = b('')

    if lldprc != 0:
        module.fail_json(msg="lldptool command failed",
                         lldprc=lldprc,
                         lldpcmd=lldpcmd,
                         system_name=system_name,
                         interface=interface,
                         stdout=lldpout,
                         stderr=lldperr,
                         start=str(startd),
                         end=str(datetime.datetime.now()),
                         delta=str(datetime.datetime.now()-startd),
                         changed=False)

    lscmd = "%s -la /sys/class/net/%s/device/" % (LS, interface)
    lsrc, lsout, lserr = module.run_command(lscmd, check_rc=True)

    if lserr is None:
        lserr = b('')

    if lsout is None:
        lsout = b('')

    if lsrc != 0:
        module.fail_json(msg="ls command failed",
                         lsrc=lsrc,
                         lscmd=lscmd,
                         system_name=system_name,
                         interface=interface,
                         stdout=lsout,
                         stderr=lserr,
                         lldpout=lldpout,
                         lldperr=lldperr,
                         start=str(startd),
                         end=str(datetime.datetime.now()),
                         delta=str(datetime.datetime.now()-startd),
                         changed=False)

    module.exit_json(lldpcmd=lldpcmd,
                     lscmd=lscmd,
                     system_name=system_name,
                     interface=interface,
                     stdout=convert_lldptool_output_to_json(interface, lldpout, lsout),
                     stderr=lldperr,
                     lldpout=lldpout,
                     lldperr=lldperr,
                     lsout=lsout,
                     lsrc=lsrc,
                     lldprc=lldprc,
                     start=str(startd),
                     end=str(datetime.datetime.now()),
                     delta=str(datetime.datetime.now()-startd),
                     changed=True)

if __name__ == '__main__':
    main()
